caikit/runtime/client/remote_module_base.py

```diff
@@ -424,54 +427,72 @@ def output_stream_parser():
-                return self._conn_channel
 
-            # Gather grpc configuration
+        def grpc_channel_construction_fn():
             target = self._get_remote_target()
             options = list(self._connection.options.items())
-
-            # Generate secure channel
-            channel = construct_grpc_channel(
+            return construct_grpc_channel(
                 target,
                 options,
                 self._tls,
                 self._connection.retries,
                 self._connection.retry_options,
             )
-            self._conn_channel = channel
-            return self._conn_channel
+
+        return self._get_remote_object(grpc_channel_construction_fn)
 
     @property
     def _http_session(self) -> Session:
         """Helper function to construct a requests Session with
         with correct credentials and TLS settings."""
-        # Short circuit if session has already been set
-        if self._conn_channel:
-            return self._conn_channel
-
-        with self._channel_lock:
-            # Check for the channel again incase it was created during lock acquisition
-            if self._conn_channel:
-                return self._conn_channel
 
-            self._conn_channel = construct_requests_session(
+        def session_construction_fn():
+            return construct_requests_session(
                 self._connection.options,
                 self._tls,
                 self._connection.timeout,
                 self._connection.retries,
                 self._connection.retry_options,
             )
-            return self._conn_channel
+
+        return self._get_remote_object(session_construction_fn)
 
     ### Generic Helper Functions
 
+    def _get_remote_object(
+        self, construction_fn: Callable[[None], Union[grpc.Channel, Session]]
+    ) -> Union[grpc.Channel, Session]:
+        """Helper function to control construction of a grpc channel or http session
+
+        Args:
+            construction_fn (Callable[[None], Union[grpc.Channel, Session]]): _description_
+
+        Returns:
+            Union[grpc.Channel, Session]: _description_
+        """
+        # If max_session_age is 0 then always return a new session/channel
+        if self._connection.max_session_age == 0:
+            return construction_fn()
+
+        with self._channel_lock:
+            # If there isn't a channel then construct a new one and set the time
+            if not self._conn_channel:
+                self._current_conn_time = datetime.now()
+                self._conn_channel = construction_fn()
+
+            # If the max session age is less then 0 then reuse the existing con channel
+            elif self._connection.max_session_age < 0:
+                pass
+
+            # If the current time is greater than the conn channel timeout then construct a
+            # new channel
+            elif datetime.now() - self._current_conn_time > self._max_conn_delta:
```

Comment:
The `elif...pass` is a bit confusing. I'd suggest merging these into a single `elif`

Reply:
Good call, I was trying to be explicit but just made it more confusing